src/use-pdc-api.ts

```diff
@@ -0,0 +1,36 @@
+import { useEffect, useState } from 'react';
+import { useOidcFetch } from '@axa-fr/react-oidc';
+import { getLogger } from './logger';
+
+const logger = getLogger('use-pdc-api');
+
+const API_URL = process.env.REACT_APP_API_URL;
+
+// Custom React hook to make authenticated requests to the configured API
+const usePdcApi = <T>(path: string): T | null => {
+  const { fetch } = useOidcFetch();
+  const [response, setResponse] = useState(null);
+
+  useEffect(() => {
+    fetch(new URL(path, API_URL))
+      .then((res) => {
+        if (!res.ok) {
+          throw new Error(`Response status: ${res.status} ${res.statusText}`);
+        }
+        return res;
+      })
+      .then((res) => res.json())
+      .then(setResponse)
+      .catch((error: unknown) => logger.error({ error }, `Error fetching ${path}`));
+
+    /* eslint-disable-next-line react-hooks/exhaustive-deps --
+     * fetch ought to be a dependency, but including it causes
+     * this effect to repeat forever
+     * https://github.com/AxaFrance/react-oidc/issues/1023
+     */
+  }, [path, setResponse]);
```

Comment:
I _think_ this implementation misuses effect dependencies.

This is treating them like function parameters, but the `useEffect` hook has access to the outer scope and doesn't need that (which is why `fetch` works). These are dependencies that should trigger the effect to run again when they change (which is why including `fetch` in the dependencies causes it to repeatedly re-run).

Assuming we only want this to run on initial render, we still want the dependency array, but I think it should be empty, right?

```suggestion
  }, []);
```

I may be totally misunderstanding what's going on here, but I don't quite follow as-is.

Reply:
Including `setResponse` is probably being over-cautious, yes. But I think we do need to include `path`, because otherwise we can't navigate without a full page reload.

I tried it by adding prev/next buttons, and removing `path` from the dependency array, and navigating to another proposal did not load any new data. I believe this is because we have the same React component tree, and the only difference React sees is the new react-router path parameter ([via the `useParams()` hook](https://reactrouter.com/en/main/hooks/use-params)). We use that to construct the API relative path, but if we do not include that relative path (the `path` function parameter here) in the dependency list, React looks and says "I don't need to re-run this effect", and so we do not update the data.

I *think* we would ideally want `fetch` to be in the dependency list, too, but I'm not completely certain. I would expect it to trigger the effect to run once its token expires and is refreshed, but that's not how it works currently (see the upstream bu g I filed), and even so that's kinda getting into IdP configuration territory - if we set very short expiration times on our tokens, do we want viewers to continuously be making API calls? Probably not, so maybe the comment is incomplete...

Reply:
Ah, excellent point about `path`, and I agree it should be passed in.

`setResponse` I'm pretty sure doesn't. I read the [React guidance here](https://legacy.reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) which helped me quite a bit, but it was really the [next section on performance](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) that demonstrated specifically that the state-setting function doesn't need to be passed in.

As for `fetch`, I assumed that `useOidcFetch()` was simply returning a custom fetch method that decorated the request with the token, and any determination about token freshness occurred when `fetch()` itself was called. But I'm at the edges of my understanding here. If you think the definition of `fetch` itself changes from render to render, then yes, let's keep it in the list (after that upstream bug is solved)! I still don't fully understand, but trust your knowledge of it more than mine.

Reply:
> `setResponse` I'm pretty sure doesn't. I read the [React guidance here](https://legacy.reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) which helped me quite a bit, but it was really the [next section on performance](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) that demonstrated specifically that the state-setting function doesn't need to be passed in, unless I'm misunderstanding our use of it.

Thank you, I hadn't read closely enough! In particular, the [old docs specifically call out `setState` as not needing to be a dependency](https://legacy.reactjs.org/docs/hooks-reference.html):

> React guarantees that `setState` function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the `useEffect` or `useCallback` dependency list.

I will fix!

> As for `fetch`, I assumed that `useOidcFetch()` was simply returning a custom fetch method that decorated the request with the token, and any determination about token freshness occurred when `fetch()` itself was called. But I'm at the edges of my understanding here. If you think the definition of `fetch` itself changes from render to render, then yes, let's keep it in the list (after that upstream bug is solved)! I still don't fully understand, but trust your knowledge of it more than mine.

Yeah, after thinking further, you're right. `useOidcFetch()` does not depend on any of the state or props we care about; it has its own internal state, but we don't need or want to re-fetch data based on any of its state - updating a token should not cause any API requests! I'll update the comment.